onnxruntime/test/optimizer/initializer_test.cc

```diff
@@ -108,28 +108,43 @@ TEST(OptimizerInitializerTest, LoadExternalData) {
 }
 
 template <typename T>
-ONNX_NAMESPACE::TensorProto_DataType GetTensorProtoDataType();
+constexpr ONNX_NAMESPACE::TensorProto_DataType GetTensorProtoDataType();
 
-#define CppTypeToTensorProto_DataType(CppType, TP_DataType)                \
-  template <>                                                              \
-  ONNX_NAMESPACE::TensorProto_DataType GetTensorProtoDataType<CppType>() { \
-    return ONNX_NAMESPACE::TP_DataType;                                    \
+#define CppTypeToTensorProto_DataType(CppType, TP_DataType)                          \
+  template <>                                                                        \
+  constexpr ONNX_NAMESPACE::TensorProto_DataType GetTensorProtoDataType<CppType>() { \
+    return ONNX_NAMESPACE::TP_DataType;                                              \
   }
 
-CppTypeToTensorProto_DataType(int8_t, TensorProto_DataType_INT8)
-CppTypeToTensorProto_DataType(uint8_t, TensorProto_DataType_UINT8)
-CppTypeToTensorProto_DataType(int32_t, TensorProto_DataType_INT32)
-CppTypeToTensorProto_DataType(int64_t, TensorProto_DataType_INT64)
-CppTypeToTensorProto_DataType(uint16_t, TensorProto_DataType_FLOAT16)
-CppTypeToTensorProto_DataType(float, TensorProto_DataType_FLOAT)
-CppTypeToTensorProto_DataType(double, TensorProto_DataType_DOUBLE)
+CppTypeToTensorProto_DataType(int8_t, TensorProto_DataType_INT8);
+CppTypeToTensorProto_DataType(uint8_t, TensorProto_DataType_UINT8);
+CppTypeToTensorProto_DataType(int32_t, TensorProto_DataType_INT32);
+CppTypeToTensorProto_DataType(int64_t, TensorProto_DataType_INT64);
+CppTypeToTensorProto_DataType(MLFloat16, TensorProto_DataType_FLOAT16);
+CppTypeToTensorProto_DataType(float, TensorProto_DataType_FLOAT);
+CppTypeToTensorProto_DataType(double, TensorProto_DataType_DOUBLE);
 
 template <typename T>
-void TestInitializerRawData() {
+std::vector<T> GetInitializerData() {
   std::vector<T> data{
       0, 1, 2, 3,
       4, 5, 6, 7,
       8, 9, 10, 11};
+  return data;
+}
+
+template <>
+std::vector<MLFloat16> GetInitializerData<MLFloat16>() {
+  std::vector<MLFloat16> data{
```

Comment:
Initialize MLFloat16 and BFloat16 with unit16 (from bits) has different meaning than doing this for other types. It sets the internal val to this value, but the real float value for MLFloat16/BFloat16 may be different. This should be OK for the UT here. But if we want to test the calculation functions (for example, add, sub, etc in the initilizer class, though I don't see any related test here), we may have trouble comparing the values between excal and expected. Using float value to generate MLFloat16 for this vector seems to be more meaningful.

Reply:
Not sure I follow. This is the original test. We actually store MLFloat16/BFloat16 as uint16 in the initialziers and Tensors, which what this test was written for. In fact, it is a requirement that the uint16 buffer casts to MLFloat16 and we use it. There was a number of occasions when this kinda code was needed to initialize it from uint16 and that is the exact meaning of the constructors for both types.